### PR TITLE
Incorporate spatial quadtree indexing.

### PR DESCRIPTION
**Pull request by feuerste**
This is required to query neighbors on the same quadtree level.

### CODE REVIEW

File: quadtree/src/lib.rs [hunk 1]
```diff
@@ -246,4 +332,13 @@ mod tests {
             assert_eq!(&NodeId::from_str(id).unwrap().to_string(), id);
         }
     }
+
```
Comment (krzesi-mir) [inline quadtree/src/lib.rs]: Let's add one more test where we explicitly test where we explicitly test the output of `SpatialNodeId::from`? For example (I might have coordinates wrong :P)
```
assert_eq!(SpatialNodeId{level: 3, x:5, y:4}, SpatialNodeId::from(NodeId::from_str("r301").unwrap());
```

File: quadtree/src/lib.rs [hunk 2]
```diff
@@ -214,9 +214,95 @@ impl fmt::Display for NodeId {
     }
 }
 
+pub enum SpatialNeighbor {
```
Comment (krzesi-mir) [inline quadtree/src/lib.rs]: nit: how about renaming this `enum` to `Direction` or `SpatialNeighborKind`?

File: quadtree/src/lib.rs [hunk 3]
```diff
@@ -214,9 +214,95 @@ impl fmt::Display for NodeId {
...
+/// on how to convert between coordinates and the quadkey.
+impl From<NodeId> for SpatialNodeId {
+    fn from(node_id: NodeId) -> Self {
+        let level = node_id.level;
+        let mut x = 0;
+        let mut y = 0;
+        for i in 1..=level {
+            let mask = 1 << (level - i);
+            match 3 & (node_id.index >> ((level - i) * 2)) {
+                1 => y |= mask,
```
Comment (krzesi-mir) [inline quadtree/src/lib.rs]: This is different than what's in the link and your coordinates are transposed w.r.t. the coordinates in the doc, in the sense that whenever they have `y` you have `x`, and whenever they have `x` you have `y` (I hope this makes sense).
Is it intentional?

File: quadtree/src/lib.rs [hunk 4]
```diff
@@ -246,4 +332,13 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_conversion() {
+        for id in &["r", "r0", "r123323"] {
+            let node_id = NodeId::from_str(id).unwrap();
+            let spatial_node_id = SpatialNodeId::from(node_id);
+            assert_eq!(NodeId::from(spatial_node_id), node_id);
+        }
+    }
```
Comment (krzesi-mir) [inline quadtree/src/lib.rs]: And maybe let's add another unit test for your `SpatialNodeId::neighbor` function?

File: quadtree/src/lib.rs (hunk 2 above)
Comment (feuerste) [reply quadtree/src/lib.rs]: Done, I renamed it to `Direction`.

File: quadtree/src/lib.rs (hunk 1 above)
Comment (feuerste) [reply quadtree/src/lib.rs]: Done, I had to switch x and y.

File: quadtree/src/lib.rs [hunk 5]
```diff
@@ -214,9 +214,95 @@ impl fmt::Display for NodeId {
...
+        let level = spatial_node_id.level;
+        let mut index = 0;
+        for i in 1..=level {
+            index <<= 2;
+            let mask = 1 << (level - i);
+            if (spatial_node_id.y & mask) != 0 {
+                index += 1;
+            }
+            if (spatial_node_id.x & mask) != 0 {
+                index += 2;
```
[the remark this answers is not in the document]
Comment (feuerste) [reply quadtree/src/lib.rs]: Done.

File: quadtree/src/lib.rs [hunk 6]
```diff
@@ -214,9 +214,95 @@ impl fmt::Display for NodeId {
...
+            let mask = 1 << (level - i);
+            match 3 & (node_id.index >> ((level - i) * 2)) {
+                1 => y |= mask,
+                2 => x |= mask,
+                3 => {
+                    x |= mask;
+                    y |= mask;
+                }
+                _ => (),
+            }
```
[the remark this answers is not in the document]
Comment (feuerste) [reply quadtree/src/lib.rs]: I changed it as proposed above.

File: quadtree/src/lib.rs (hunk 3 above)
Comment (feuerste) [reply quadtree/src/lib.rs]: Actually we here have a right hand coordinate frame (in contrast to the left hand frame for the Bing link), so I had to switch x and y.

File: quadtree/src/lib.rs [hunk 7]
```diff
@@ -214,9 +214,95 @@ impl fmt::Display for NodeId {
...
+/// See e.g. https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system
+/// on how to convert between coordinates and the quadkey.
+impl From<NodeId> for SpatialNodeId {
+    fn from(node_id: NodeId) -> Self {
+        let level = node_id.level;
+        let mut x = 0;
+        let mut y = 0;
+        for i in 1..=level {
+            let mask = 1 << (level - i);
+            match 3 & (node_id.index >> ((level - i) * 2)) {
```
[the remark this answers is not in the document]
Comment (feuerste) [reply quadtree/src/lib.rs]: Done.

File: quadtree/src/lib.rs (hunk 4 above)
Comment (feuerste) [reply quadtree/src/lib.rs]: Puh, were should I start and end here? I think the function is trivial enough not to test it...

File: quadtree/src/lib.rs (hunk 3 above)
Comment (krzesi-mir) [reply quadtree/src/lib.rs]: :+1:

File: quadtree/src/lib.rs (hunk 4 above)
Comment (krzesi-mir) [reply quadtree/src/lib.rs]: yeah you are probably right :P

**Comment by feuerste** [comment]
Thanks for your review!

**Pull request merged**